src/node_report_module.cc

```diff
@@ -0,0 +1,315 @@
+  fflush(stderr);
+  raise(SIGABRT);
+}
+
+void OnUncaughtException(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  std::string cmdline = node::per_process_opts->cmdline;
+  std::shared_ptr<PerIsolateOptions> options = env->isolate_data()->options();
+
+  // Trigger report if requested
+  if (options->report_catch_exception) {
+    TriggerNodeReport(isolate, env, "exception", __func__, nullptr, info[0]);
+  }
+}
+
+/*******************************************************************************
+ * Signal handler for report action, called from JS land (util.js)
+ ******************************************************************************/
+
+void OnUserSignal(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  CHECK(info[0]->IsString());
+  Local<String> str = info[0].As<String>();
+  String::Utf8Value value(isolate, str);
+  TriggerNodeReport(
+      isolate,
+      env,
+      *value,
+      __func__,
+      nullptr,
+      MaybeLocal<Value>());
+}
+
+/*******************************************************************************
+ * Native module initializer function, called when the module is require'd
+ ******************************************************************************/
+void InitializeNodeReport(Isolate* isolate, Environment* env) {
+  // Register the boot time of the process, for
+  // computing resource consumption average etc.
+  // std::string options;
+  IsolateData* id = env->isolate_data();
+  std::shared_ptr<PerIsolateOptions> options = env->isolate_data()->options();
+  time(&boot_time);
+  if (options->report_catch_fatalerror && id->error_hook_initialised == false) {
+    env->isolate()->SetFatalErrorHandler(OnFatalError);
+    id->error_hook_initialised = true;
+  }
+
+  if (options->report_signal == "") options->report_signal = "SIGUSR2";
+}
+// A method to sync up data elements in the JS land with its
+// corresponding elements in the C++ world. Required because
+// (i) the tunables are first intercepted through the CLI but
+// later modified via APIs. (ii) the report generation events
+// are controlled partly from C++ and partly from JS.
+void SyncConfig(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Local<Context> context = env->context();
+  std::shared_ptr<PerIsolateOptions> options = env->isolate_data()->options();
+  IsolateData* id = env->isolate_data();
+
+  CHECK_EQ(info.Length(), 2);
+  Local<Object> obj = info[0]->ToObject(context).ToLocalChecked();
+  bool sync = info[1].As<Boolean>()->Value();
+
+  // events array
+  Local<String> eventskey = FIXED_ONE_BYTE_STRING(env->isolate(), "events");
+  Local<Value> events_unchecked =
+      obj->Get(context, eventskey).ToLocalChecked();
```

Comment:
Can you handle exceptions and return back to JS land if they occur?

Reply:
- should I throw, or return NULL, or return an error?
 - is it required for every object access (get|set) that is attempted, or only this site? It may bloat up the code? Given that the caller is always trusted (`./lib`) with call sites well defined, may be could relax the error check a bit? I am also not seeing such checks around the object access, in the code base, can you please point to a reference?

Reply:
> should I throw, or return NULL, or return an error?

We generally don’t use C++ exceptions, if that’s what you’re asking. Simply `return;` should be the right thing to do here if the result of `Get()` is empty – an empty `Maybe<>` or `MaybeLocal<>` is used by the V8 API to indicate that an exception is scheduled, so the calling JS code will see the exception as if it was thrown from this method.

> is it required for every object access (get|set) that is attempted, or only this site?

I’m talking about every site where the V8 API uses `MaybeLocal`s, yeah…

> It may bloat up the code?

Unfortunately, yes.

> Given that the caller is always trusted (`./lib`) with call sites well defined, may be could relax the error check a bit?

So … the story is a bit complicated. We can use `ToLocalChecked()` if we know exactly that is is not going to call any JS code. In this case, this could at least theoretically happen if somebody messes with `Object.prototype` – it’s obviously not something we should encourage, but I’d personally prefer for new code to be written in a way that is resilient against any JS manipulation.

(And in the presence of `worker.terminate()`, *any* JS code can fail, not just code that throws exceptions on its own – auditing every usage of `ToLocalChecked()` & friends in our codebase is part of the to-do list for Workers.)

> I am also not seeing such checks around the object access, in the code base, can you please point to a reference?

I guess this would be a good example of how the code can still be somewhat compact?

https://github.com/nodejs/node/blob/fd6a7d4db185db6c0084b6a593d698568f0ffe9f/src/node.cc#L1251-L1258

The fact that you’re not seeing a lot of this is mostly because

Reply:
fixed, ptal. thanks!